File: python/2296.py
class TextEditor:

    def __init__(self):
        self.text = ''
        self.cursor = 0
        

    def addText(self, text: str) -> None:
        self.text = self.text[:self.cursor] + text + self.text[self.cursor:]
        self.cursor += len(text)
        

    def deleteText(self, k: int) -> int:
        if self.cursor-k < 0:
            k = self.cursor
        self.text = self.text[0:self.cursor-k] + self.text[self.cursor:]
        self.cursor -= k
        return k

    def cursorLeft(self, k: int) -> str:
        if self.cursor-k < 0:
            k = self.cursor
        self.cursor -= k
        return self.getLeft10Text()
        

    def cursorRight(self, k: int) -> str:
        if self.cursor+k > len(self.text):
            k = len(self.text) - self.cursor
        self.cursor += k
        return self.getLeft10Text()
    
    def getLeft10Text(self):
        if self.cursor-10 < 0:
            return self.text[0:self.cursor]
        return self.text[self.cursor-10:self.cursor]


#["TextEditor","addText","deleteText","addText","cursorRight","cursorLeft","deleteText","cursorLeft","cursorRight"]
#[[],["leetcode"],[4],["practice"],[3],[8],[10],[2],[6]]
obj = TextEditor()
obj.addText("leetcode")
print(obj.deleteText(4))
obj.addText("practice")
print(obj.cursorRight(3))
print(obj.cursorLeft(8))
print(obj.deleteText(10))
print(obj.cursorLeft(2))
print(obj.cursorRight(6))
print('\n'+obj.text)
print(obj.cursor)

#["TextEditor","addText","cursorLeft","cursorLeft","addText","deleteText"]
#[[],["jxarid"],[5],[10],["du"],[20]]

obj = TextEditor()
obj.addText("jxarid")
print(obj.cursorLeft(5))
print(obj.cursorLeft(10))
obj.addText("du")
print(obj.deleteText(20))
print('\n'+obj.text)
print(obj.cursor)

File: js/2296.js
var Node = function(val) {
    this.val = val;
    this.prev = null;
    this.next = null;
};

Node.prototype.insert = function(val) {
    var node = new Node(val);
    node.next = this;
    node.prev = this.prev;
    if (this.prev) {
        this.prev.next = node;
    }
    this.prev = node;
};

Node.prototype.remove = function() {
    var node = this.prev;
    this.prev = node.prev;
    if (node.prev) {
        node.prev.next = this;
    }
};

Node.prototype.range = function(end) {
    var result = '';
    var node = this;
    while (node !== end) {
        result += node.val;
        node = node.next;
    }
    return result;
};

var TextEditor = function() {
    this.cursor = new Node('\0'); // 初始光标节点
};

TextEditor.prototype.addText = function(text) {
    for (let i = 0; i < text.length; i++) {
        this.cursor.insert(text[i]);
    }
};

TextEditor.prototype.deleteText = function(k) {
    var count = 0;
    while (k > 0 && this.cursor.prev) {
        this.cursor.remove();
        k--;
        count++;
    }
    return count;
};

TextEditor.prototype.cursorLeft = function(k) {
    while (k > 0 && this.cursor.prev) {
        this.cursor = this.cursor.prev;
        k--;
    }
    var head = this.cursor;
    for (let i = 0; i < 10 && head.prev; i++) {
        head = head.prev;
    }
    return head.range(this.cursor);
};

TextEditor.prototype.cursorRight = function(k) {
    while (k > 0 && this.cursor.next) {
        this.cursor = this.cursor.next;
        k--;
    }
    var head = this.cursor;
    for (let i = 0; i < 10 && head.prev; i++) {
        head = head.prev;
    }
    return head.range(this.cursor);
};

作者：力扣官方题解
链接：https://leetcode.cn/problems/design-a-text-editor/solutions/3078930/she-ji-yi-ge-wen-ben-bian-ji-qi-by-leetc-o4by/
来源：力扣（LeetCode）
著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。